components/script/dom/mutationobserver.rs

```diff
@@ -37,4 +69,150 @@ impl MutationObserver {
     }
+
+    /// https://dom.spec.whatwg.org/#queue-a-mutation-observer-compound-microtask
+    /// Queue a Mutation Observer compound Microtask.
+    pub fn queueMutationObserverCompoundMicrotask(compoundMicrotask: Microtask) {
+        // Step 1
+        if ScriptThread::get_mutation_observer_compound_microtask_queued() {
+            return;
+        }
+        // Step 2
+        ScriptThread::set_mutation_observer_compound_microtask_queued(true);
+        // Step 3
+        ScriptThread::enqueue_microtask(compoundMicrotask);
+    }
+
+    /// https://dom.spec.whatwg.org/#notify-mutation-observers
+    /// Notify Mutation Observers.
+    pub fn notifyMutationObservers() {
+        // Step 1
+        ScriptThread::set_mutation_observer_compound_microtask_queued(false);
+        // Step 2
+        let notifyList = ScriptThread::get_mutation_observer();
+        // Step 3, Step 4 not needed as Servo doesn't implement anything related to slots yet.
+        // Step 5: Ignore the specific text about execute a compound microtask.
+        // Step 6 not needed as Servo doesn't implement anything related to slots yet.
+    }
+
+    //https://dom.spec.whatwg.org/#queueing-a-mutation-record
+    //Queuing a mutation record
+    pub fn queue_a_mutation_record(target: &Node, attr_type: Mutation) {
+        let mut given_name = "";
+        let mut given_namespace = "";
+        let mut old_value = "";
+        // Step 1
+        let mut interestedObservers: Vec<MutationObserver> = vec![];
+        let mut pairedStrings: Vec<DOMString> = vec![];
+        // Step 2
+        let mut nodes: Vec<Root<Node>> = vec![];
+        for ancestor in target.inclusive_ancestors() {
+            nodes.push(ancestor);
+        }
+        // Step 3
+        for node in &nodes {
+            for registered_observer in node.registered_mutation_observers_for_type().borrow().iter() {
+                match attr_type {
+                    Mutation::Attribute { ref name, ref namespace, ref oldValue, ref newValue } => {
+                        given_name = &*name.clone();
+                        given_namespace = &*namespace.clone();
+                        old_value = &*oldValue.clone();
+                    },
+                }
+                let condition1: bool = node != &Root::from_ref(target) &&
+                    !registered_observer.options.into_inner().subtree;
+                let condition2: bool = given_name == "attribute" &&
+                    registered_observer.options.into_inner().attributes == Some(false);
+                let condition3: bool = (given_name == "attribute" &&
+                    registered_observer.options.into_inner().attributeFilter != None) &&
+                    (given_namespace != "");
+                let condition4: bool = given_name == "characterData" &&
+                    registered_observer.options.into_inner().characterData == Some(false);
+                let condition5: bool = given_name == "childList" &&
```

Comment:
Why are we comparing `given_name` here? Since we only support mutations on attributes, we only need to care about specification text that refers to the "type" variable being "attributes".